### PR TITLE
Fix NullPointerException on null partition insert

### LINKED ISSUE #1735: [FEA] Fix NPE for static partition with null in GpuInsertIntoHadoopFsRelationCommand
**Is your feature request related to a problem? Please describe.**
Update GpuInsertIntoHadoopFsRelationCommand for [SPARK-34223][SQL] FIX NPE for static partition with null in  InsertIntoHadoopFsRelationCommand

We have copied  InsertIntoHadoopFsRelationCommand into  GpuInsertIntoHadoopFsRelationCommand so we need to apply similar fix: https://github.com/apache/spark/commit/f2ab524fca

### PR DESCRIPTION
**Pull request by gerashegalov**
Port https://github.com/apache/spark/pull/31320 to close #1735

Signed-off-by: Gera Shegalov <gera@apache.org>

### CODE REVIEW

**Comment by gerashegalov** [comment]
build

**Review by jlowe — CHANGES_REQUESTED** [review]
#1735 is targeting the 0.4 release yet this is targeting branch-0.5.

**Comment by gerashegalov** [comment]
> #1735 is targeting the 0.4 release yet this is targeting branch-0.5.

I thought we merge to the latest branch first and then cherry-pick; changing

**Comment by gerashegalov** [comment]
build

**Review by jlowe — DISMISSED** [review]

**Comment by gerashegalov** [comment]
:eyes: into test failuers

**Comment by gerashegalov** [comment]
build

File: pom.xml
Comment (jlowe) [inline pom.xml:189]: This PR is suddenly doing something much more than just fixing an NPE.  It's updating the build to pull 3.1.1 instead of a snapshot, arguably a separate PR.

File: pom.xml
Comment (jlowe) [inline pom.xml:145]: Is this an issue with the public snapshots or the snapshots from URM?  If it was passing for you locally using the Apache snapshots but failing in CI then it's likely an issue with URM snapshots and we should just fix the URM snapshot build.

File: pom.xml
Comment (jlowe) [inline pom.xml:149]: I'm not thrilled putting this in.  If this is expected to be very temporary, why not just hold off merging this PR until 3.1.1 is officially released or the snapshot build is fixed?

File: tests/pom.xml
Comment (jlowe) [inline tests/pom.xml:80]: Indentation is off here.

**Review by jlowe — CHANGES_REQUESTED** [review]

File: pom.xml
Comment (gerashegalov) [reply pom.xml:145]: it's the "issue" with public spark snapshots

File: pom.xml
Comment (gerashegalov) [reply pom.xml:189]: surely it can be a separate PR, but it was definitely a prerequisite for passing the test locally

Comment (jlowe) [reply pom.xml:189]: No doubt.  However this PR is now submarining a much more impactful change than the headline implies.   One of my pet peeves is PRs with a headline like "fix some minor test thing" that ends up changing way more than that.

File: pom.xml
Comment (gerashegalov) [reply pom.xml:149]: ok, I we can make it a property or environment variable passed from outside, which is actually convenient now that we are waiting for rc3 for 3.1.1
e.g.  this would fail
```
$ find ~/.m2/repository -path \*/spark\*/3.1.1 -type d | xargs rm -r
$ mvn -Pspark311tests,snapshot-shims test -Dspark.rc.repo=https://repository.apache.org/content/repositories/orgapachespark-1364 -Dcuda.version=cuda11 -DwildcardSuites=com.nvidia.spark.rapids.InsertPartition311Suite
```

this would succeed
```
$ find ~/.m2/repository -path \*/spark\*/3.1.1 -type d | xargs rm -r
$ mvn -Pspark311tests,snapshot-shims test -Dspark.rc.repo=https://repository.apache.org/content/repositories/orgapachespark-1365 -Dcuda.version=cuda11 -DwildcardSuites=com.nvidia.spark.rapids.InsertPartition311Suite
```

Comment (jlowe) [reply pom.xml:149]: But then we'd have to update all the CI/CD scripts or they will all break.

I think we should just wait until 3.1.1 gets sorted out before merging this change.  We can't say 0.4 supports 3.1 until 3.1 is out anyway. :smile:  In the end I think it's a lot less code churn and less risk.

**Comment by jlowe** [comment]
The 3.1.1 shim has been updated and now Spark 3.1.1 is using the official artifacts.  @gerashegalov if you upmerge this PR to the latest on branch-0.4 then it should run with 3.1.1 and the tests should pass.

**Comment by gerashegalov** [comment]
build

**Comment by gerashegalov** [comment]
> The 3.1.1 shim has been updated and now Spark 3.1.1 is using the official artifacts. @gerashegalov if you upmerge this PR to the latest on branch-0.4 then it should run with 3.1.1 and the tests should pass.

@jlowe 
:heavy_check_mark: merged branch-0.4 into the PR branch and `mvn -Pspark311tests clean test -Dcuda.version=cuda11 -DwildcardSuites=com.nvidia.spark.rapids.InsertPartition311Suite ` passes

**Review by jlowe — APPROVED** [review]

**Pull request merged**
